inc/saitunnel.h

```diff
@@ -0,0 +1,444 @@
+
+    /* -- */
+    /* Custom range base value */
+    SAI_TUNNEL_MAP_ATTR_CUSTOM_RANGE_BASE = 0x10000000
+
+} sai_tunnel_map_attr_t;
+
+/**
+ * @brief Create tunnel Map
+ *
+ * @param[out] tunnel_map_id tunnel Map Id
+ * @param[in] attr_count number of attributes
+ * @param[in] attr_list array of attributes
+ *
+ * @return  SAI_STATUS_SUCCESS on success
+ *          Failure status code on error
+ */
+typedef sai_status_t(*sai_create_tunnel_map_fn)(
+    _Out_ sai_object_id_t* tunnel_map_id,
+    _In_ uint32_t attr_count,
+    _In_ const sai_attribute_t *attr_list
+    );
+
+/**
+ * @brief Remove tunnel Map
+ *
+ *  @param[in] tunnel_map_id tunnel Map id to be removed.
+ *
+ *  @return  SAI_STATUS_SUCCESS on success
+ *           Failure status code on error
+ */
+typedef sai_status_t(*sai_remove_tunnel_map_fn) (
+    _In_  sai_object_id_t   tunnel_map_id
+    );
+
+/**
+ * @brief Set attributes for tunnel map
+ *
+ * @param[in] tunnel_map_id tunnel Map Id
+ * @param[in] attr attribute to set
+ *
+ * @return  SAI_STATUS_SUCCESS on success
+ *          Failure status code on error
+ */
+typedef sai_status_t(*sai_set_tunnel_map_attribute_fn)(
+    _In_ sai_object_id_t  tunnel_map_id,
+    _In_ const sai_attribute_t *attr
+    );
+
+/**
+ * @brief  Get attrbutes of tunnel map
+ *
+ * @param[in] tunnel_map_id  tunnel map id
+ * @param[in] attr_count  number of attributes
+ * @param[inout] attr_list  array of attributes
+ *
+ * @return SAI_STATUS_SUCCESS on success
+ *        Failure status code on error
+ */
+typedef sai_status_t(*sai_get_tunnel_map_attribute_fn)(
+    _In_ sai_object_id_t   tunnel_map_id,
+    _In_ uint32_t attr_count,
+    _Inout_ sai_attribute_t *attr_list
+    );
+
+typedef enum _sai_tunnel_type_t
+{
+    SAI_TUNNEL_IPINIP,
+
+    SAI_TUNNEL_IPINIP_GRE,
+
+    SAI_TUNNEL_IPINIP_GRE_AND_KEY,
+
+    SAI_TUNNEL_VXLAN,
+
+    SAI_TUNNEL_MPLS,
+
+} sai_tunnel_type_t;
+
+typedef enum _sai_tunnel_ttl_mode_t
+{
+    SAI_TUNNEL_TTL_COPY_FROM_INNER,
+
+    SAI_TUNNEL_TTL_USER_DEFINE
+
+} sai_tunnel_ttl_mode_t;
+
+typedef enum _sai_tunnel_dscp_mode_t
+{
+    SAI_TUNNEL_DSCP_COPY_FROM_INNER,
+
+    SAI_TUNNEL_DSCP_USER_DEFINE
+
+} sai_tunnel_dscp_mode_t;
+
+typedef enum _sai_tunnel_attr_t
+{
+    /** READ-WRITE */
+
+    /** tunnel type [sai_tunnel_type_t] (MANDATORY_ON_CREATE|CREATE_ONLY) */
+    SAI_TUNNEL_ATTR_TYPE,
+
+    /** tunnel underlay interface [sai_object_id_t] */
+    SAI_TUNNEL_ATTR_UNDERLAY_INTERFACE,
+
+    /** tunnel overlay interafce [sai_object_id_t] */
+    SAI_TUNNEL_ATTR_OVERLAY_INTERFACE,
+
+    /** tunnel encap attribute*/
+
+    /** tunnel ip verssion ipv4/ipv6 [sai_ip_addr_family_t] 
+     *  (MANDATORY_ON_CREATE when (CREATE_ONLY) */
+    SAI_TUNNEL_ATTR_IP_VER,
+
+    /** tunnel src ip [sai_ip_address_t] (MANDATORY_ON_CREATE when 
+     *  SAI_TUNNEL_ATTR_TYPE=SAI_TUNNEL_IPINIP,SAI_TUNNEL_IPINIP_GRE,
+     *  SAI_TUNNEL_IPINIP_GRE_AND_KEY) (CREATE_ONLY) */
+    SAI_TUNNEL_ATTR_SRC_IP,
```

Comment:
Since sai_tunnel_attr_t is used for both encap and decap, SAI_TUNNEL_ATTR_SRC_IP might not be a good name as the IP is the src IP for the encap, but the dest IP for the decap. Would it be better to rename it to SAI_TUNNEL_ATTR_IP?